Adding ConfigCat source documentation (#6596)
segmentio/segment-docs: src/connections/sources/catalog/cloud-apps/configcat/index.md

> ---
> title: ConfigCat Source
> ---

Reviewer: ```suggestion
---
title: ConfigCat Source
id: nEjnxv4kbB
beta: true
---
```

        - `userId` (optional): the user object's identifier used during feature flag evaluation
        - `user` (optional): the user object used during feature flag evaluation.
    - You can find code samples in the [ConfigCat Segment Source documentation](https://configcat.com/docs/integrations/segment/#analytics){:target="_blank”}.

## Stream

ConfigCat uses our stream Source component to send Segment event data. It uses a `track` method to send data to Segment. These events are then available in any destination that accepts events, and available in a schema in your data warehouse, so you can query using SQL.

The default behavior is for ConfigCat to pass the user object's identifier property used during the feature flag evaluation as the userId or sending the ConfigCat User's ID who made the changes as the userId.

## Events

The table below lists events that ConfigCat sends to Segment. These events appear as tables in your warehouse, and as regular events in other Destinations. ConfigCat includes the `userId` if available.

| Event Name           | Description            |
| ------------------   | ---------------------- |

> | featureFlagEvaluated | Feature flag evaluated |
> | featureFlagChanged   | Feature flag changed   |

Reviewer: ```suggestion
| Feature Flag Evaluated | Feature flag evaluated |
| Feature Flag Changed   | Feature flag changed   |
```

Could you provide just a bit more of a description for what these events are? 

Additionally, Looking at the events in your [source's debugger](https://app.segment.com/configcat/sources/configcat/debugger), I am seeing that you are using `camelCasing` for both event names and event properties. [We actually recommend](https://segment.com/academy/collecting-data/naming-conventions-for-clean-data/) that this is configured to use `Proper Case` for event names, and `snake_case` for event properties. 

Could you update the naming conventions for your source's event payloads? If not, could you please provide some details as to why?

## Event Properties

The tables below list the properties included in the events listed above.

### `featureFlagEvaluated` event

| Property Name    | Description                                                      |
| ---------------- | ---------------------------------------------------------------- |

> | `featureFlagKey` | The feature flag's key                                           |
> | `value`          | The evaluated feature flag's value or Variation ID               |
> | `variationId`    | The evaluated feature flag's Variation ID                        |
> | `userId`         | The user object's identifier used during feature flag evaluation |
> | `user`           | The user object used during feature flag evaluation              |

Reviewer: ```suggestion
| `feature_flag_key` | The feature flag's key                                           |
| `value`          | The evaluated feature flag's value or Variation ID               |
| `variation_id`    | The evaluated feature flag's Variation ID                        |
| `user`           | The user object used during feature flag evaluation              |
```
I noticed that you included `userId` within the properties object. Within the Segment spec, this is a key property that is only to be placed at the top level of the payload. So please ensure that you have removed `userId` or `anonymousId` from the properties object. 

[We actually recommend](https://segment.com/academy/collecting-data/naming-conventions-for-clean-data/) that this is configured to use `Proper Case` for event names, and `snake_case` for event properties.

Could you update the naming conventions for your source's event payloads? If not, could you please provide some details as to why?

### `featureFlagChanged` event

| Property Name   | Description                                                        |
| ----------------- | ---------------------------------------------------------------- |

> | `details`         | A brief summary of the change                                    |
> | `timestamp`       | When the change happened                                         |
> | `userId`          | The ConfigCat user's ID who made the changes                     |
> | `userEmail`       | The ConfigCat user's email who made the changes                  |
> | `userFullName`    | The ConfigCat user's full name who made the changes              |
> | `userId`          | The ConfigCat user's ID Who made the changes                     |
> | `productId`       | The product's ID where the change happened                       |
> | `productName`     | The product's name where the change happened                     |
> | `configId`        | The config's ID where the change happened                        |
> | `configName`      | The config's name where the change happened                      |
> | `environmentId`   | The environment's ID where the change happened                   |
> | `environmentName` | The environment's name where the change happened                 |
> | `url`             | A direct link to the config/feature flag                         |
> | `userId`          | The user object's identifier used during feature flag evaluation |
> | `user`            | The user object used during feature flag evaluation              |

Reviewer: ```suggestion
| `details`         | A brief summary of the change                                    |
| `user_email`       | The ConfigCat user's email who made the changes                  |
| `user_full_name`    | The ConfigCat user's full name who made the changes              |
| `product_id`       | The product's ID where the change happened                       |
| `product_name`     | The product's name where the change happened                     |
| `config_id`        | The config's ID where the change happened                        |
| `config_name`      | The config's name where the change happened                      |
| `environment_id`   | The environment's ID where the change happened                   |
| `environment_name` | The environment's name where the change happened                 |
| `url`             | A direct link to the config/feature flag                         |
| `user`            | The user object used during feature flag evaluation              |
```

I noticed that you included `userId` and `timestamp` within the properties object. Within the Segment spec, these are key properties that are only to be placed at the top level of the payload. So please ensure that you have removed `userId` or `anonymousId` and `timestamp` from the properties object. 

[We actually recommend](https://segment.com/academy/collecting-data/naming-conventions-for-clean-data/) that this is configured to use `Proper Case` for event names, and `snake_case` for event properties.

Could you update the naming conventions for your source's event payloads? If not, could you please provide some details as to why?

    - **ConfigCat SDK:** Initialize with your ConfigCat SDK key.
    - **Segment SDK:** Set up with the Segment Write Key acquired while adding the ConfigCat source in Segment.
3. **Integrate Feature Flag Evaluations:**
    - During the initialization of the ConfigCat SDK, subscribe to the `flagEvaluated` hook.
    - Send feature flag evaluation data to Segment using the `Feature Flag Evaluated` event name. Include the following parameters:
        - `feature_flag_key`: the feature flag's key
        - `value`: the evaluated feature flag's value or Variation ID
        - `variation_id` (optional): the evaluated feature flag's Variation ID
        - `user` (optional): the user object used during feature flag evaluation.
    - You can find code samples in the [ConfigCat Segment Source documentation](https://configcat.com/docs/integrations/segment/#analytics){:target="_blank”}.

## Stream

ConfigCat uses our stream Source component to send Segment event data. It uses a `track` method to send data to Segment. These events are then available in any destination that accepts events, and available in a schema in your data warehouse, so you can query using SQL.

> The default behavior is for ConfigCat to pass the user object's identifier property used during the feature flag evaluation as the userId or sending the ConfigCat User's ID who made the changes as the userId.

Reviewer: What happens when there is no user object identifier? Is the default to send `ConfigCat User's ID`as the Segment userId in it's place? 

Will there every be a situation where nothing is used as the Segment `userId`, and you send an event with only with `anonymousId`? If so what is used for `anonymousId`

Author: There are two ways how we are sending data to Segment. They are totally different and they have nothing in common related to User ID handling. 

The feature flag evaluated event is sent from ConfigCat's customer's application with a code snippet we provide in our documentation. The user object's identifier is our customer's application's user's identifier. If that is not provided we are sending an empty string. However it seems to be pointless at first glance, but in ConfigCat terminology it totally makes sense to not have an user object during feature flag evaluation. But we would like to send events in this case too. We don't send explicitly anything in this case as anonymusId, but the code snippets we provide are using Segment's SDKs. Is it handled there? What is your suggestion in this case? Should we generate a GUID and send it as the anonymusid? These applications don't know anything about `ConfigCat User's ID`.

The feature flag changed event is sent from ConfigCat's backend and we always provide the userId there (`ConfigCat User's ID`).

Reviewer: thanks for clarifying. no need for any changes, just need to change the copy in the documentation. please see my suggestion below.

> The default behavior is for ConfigCat to pass the user object's identifier property used during the feature flag evaluation as the userId or sending the ConfigCat User's ID who made the changes as the userId.

Reviewer: ```suggestion
In the "Feature Flag Evaluation" event, the default behavior for ConfigCat is to use the user-object's identifier property as the Segment `userId`. If no identifier is present, the event will be sent with a randomly generated `anonymousId`.

For "Feature Flag Changed" events, the ConfigCat User's ID who made the changes will be sent as the Segment userId.
```

Reviewer: @laliconfigcat thanks for clarifying. Could you please update to the following^

## Events

> The table below lists events that ConfigCat sends to Segment. These events appear as tables in your warehouse, and as regular events in other Destinations. ConfigCat includes the `userId` if available.

Reviewer: ```suggestion
The table below lists events that ConfigCat sends to Segment. These events appear as tables in your warehouse, and as regular events in other Destinations.
```

"ConfigCat includes the `userId` if available." can be removed as this will be clarified in the section above

Reviewer: @laliconfigcat 
could you make this update as well^

    - During the initialization of the ConfigCat SDK, subscribe to the `flagEvaluated` hook.
    - Send feature flag evaluation data to Segment using the `Feature Flag Evaluated` event name. Include the following parameters:
        - `feature_flag_key`: the feature flag's key
        - `value`: the evaluated feature flag's value or Variation ID
        - `variation_id` (optional): the evaluated feature flag's Variation ID
        - `user` (optional): the user object used during feature flag evaluation.
    - You can find code samples in the [ConfigCat Segment Source documentation](https://configcat.com/docs/integrations/segment/#analytics){:target="_blank”}.

## Stream

ConfigCat uses our stream Source component to send Segment event data. It uses a `track` method to send data to Segment. These events are then available in any destination that accepts events, and available in a schema in your data warehouse, so you can query using SQL.

In the "Feature Flag Evaluation" event, the default behavior for ConfigCat is to use the user-object's identifier property as the Segment `userId`. If no identifier is present, the event will be sent with a randomly generated `anonymousId`.

> For "Feature Flag Changed" events, the ConfigCat User's ID who made the changes will be sent as the Segment userId.

Reviewer: ```suggestion
For "Feature Flag Changed" events, the ConfigCat User's ID who made the changes will be sent as the Segment `userId`.
```
